engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,848 @@
+	pendings    sync.WaitGroup
+	logger      *zap.Logger
+
+	kvClient   metaModel.KVClient
+	tableAgent TableAgent
+	jobID      string
+	jobStore   *metadata.JobStore
+}
+
+// NewDDLCoordinator creates a new DDLCoordinator.
+func NewDDLCoordinator(jobID string, kvClient metaModel.KVClient, tableAgent TableAgent, jobStore *metadata.JobStore, pLogger *zap.Logger) *DDLCoordinator {
+	return &DDLCoordinator{
+		tableAgent:  tableAgent,
+		tables:      make(map[metadata.TargetTable]map[metadata.SourceTable]struct{}),
+		jobID:       jobID,
+		kvClient:    kvClient,
+		shardGroups: make(map[metadata.TargetTable]*shardGroup),
+		jobStore:    jobStore,
+		logger:      pLogger.With(zap.String("component", "ddl_coordinator")),
+	}
+}
+
+// ClearMetadata clears metadata.
+func (c *DDLCoordinator) ClearMetadata(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// wait all pending ddls coordinated.
+	c.pendings.Wait()
+
+	return metadata.DelAllDroppedColumns(ctx, c.kvClient)
+}
+
+// Reset resets the ddl coordinator.
+func (c *DDLCoordinator) Reset(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// wait all pending ddls coordinated.
+	c.pendings.Wait()
+
+	jobCfg, err := c.jobStore.GetJobCfg(ctx)
+	if err != nil {
+		return err
+	}
+	if jobCfg.ShardMode == "" {
+		c.logger.Info("non-shard-mode, skip reset")
+		return nil
+	}
+	c.logger.Info("reset ddl coordinator")
+
+	// fetch all tables which need to be coordinated.
+	tables, err := c.tableAgent.FetchAllDoTables(ctx, jobCfg)
+	if err != nil {
+		return err
+	}
+
+	for targetTable, sourceTables := range tables {
+		c.tables[targetTable] = make(map[metadata.SourceTable]struct{}, 0)
+		for _, sourceTable := range sourceTables {
+			c.tables[targetTable][sourceTable] = struct{}{}
+		}
+	}
+	return nil
+}
+
+// Coordinate coordinates ddls.
+func (c *DDLCoordinator) Coordinate(ctx context.Context, item *metadata.DDLItem) ([]string, optimism.ConflictStage, error) {
+	c.mu.Lock()
+	c.pendings.Add(1)
+	defer c.pendings.Done()
+	c.mu.Unlock()
+
+	jobCfg, err := c.jobStore.GetJobCfg(ctx)
+	if err != nil {
+		return nil, optimism.ConflictError, err
+	} else if jobCfg.ShardMode == "" {
+		return nil, optimism.ConflictError, errors.New("coordinate error with non-shard-mode")
+	}
+
+	for {
+		// create shard group if not exists.
+		g, err := c.loadOrCreateShardGroup(ctx, item.TargetTable, jobCfg)
+		if err != nil {
+			return nil, optimism.ConflictError, err
+		}
+
+		// each group only processes one item at a time
+		g.mu.Lock()
+		// if the group is deleted by another DDLItem, we should recreate it.
+		if g.deleted {
+			g.mu.Unlock()
+			continue
+		}
+		ddls, conflictStage, err := g.handle(ctx, item)
+		if err != nil {
+			g.mu.Unlock()
+			return ddls, conflictStage, err
+		}
+
+		g.deleted = g.isResolved(ctx)
+		// if all source table is deleted or the shard group is resolved, we should remove the shard group.
+		if g.deleted {
+			c.removeShardGroup(ctx, item.TargetTable)
```

Comment:
seems a dead lock, here we acquired g.mu then c.mu, but in ShowDDLLocks we reverted

Reply:
[7622504](https://github.com/pingcap/tiflow/pull/7523/commits/76225045c73a09d8d566c3938f2c07119395e8a8)